cmd/libsnap-confine-private/cgroup-support-test.c

```diff
@@ -0,0 +1,284 @@
+    }
+    g_test_trap_subprocess(NULL, 0, 0);
+    g_test_trap_assert_failed();
+    g_test_trap_assert_stderr("cannot open directory entry \"badperm\": Permission denied\n");
+}
+
+static void sc_set_self_cgroup_path(const char *mock) { self_cgroup = mock; }
+
+typedef struct _cgroupv2_own_group_fixture {
+    char *p;
+} cgroupv2_own_group_fixture;
+
+static void cgroupv2_own_group_set_up(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+    int fd = g_file_open_tmp("s-c-unit-test.XXXXXX", &fixture->p, &err);
+    g_assert_no_error(err);
+    g_close(fd, &err);
+    g_assert_no_error(err);
+    sc_set_self_cgroup_path(fixture->p);
+}
+
+static void cgroupv2_own_group_tear_down(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    sc_set_self_cgroup_path("/proc/self/cgroup");
+    g_remove(fixture->p);
+    g_free(fixture->p);
+}
+
+static void test_sc_cgroupv2_own_group_path_simple_happy(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+    g_autofree const char *p = NULL;
+    g_file_set_contents(fixture->p, (char *)user_data, -1, &err);
+    g_assert_no_error(err);
+    p = sc_cgroup_v2_own_path_full();
+    g_assert_cmpstr(p, ==, "/foo/bar/baz.slice");
+}
+
+static void test_sc_cgroupv2_own_group_path_empty(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+    g_autofree const char *p = NULL;
+    g_file_set_contents(fixture->p, (char *)user_data, -1, &err);
+    g_assert_no_error(err);
+    p = sc_cgroup_v2_own_path_full();
+    g_assert_null(p);
+}
+
+static void _test_sc_cgroupv2_own_group_path_die_with_message(const char *msg) {
+    if (g_test_subprocess()) {
+        g_autofree const char *p = NULL;
+        p = sc_cgroup_v2_own_path_full();
+    }
+    g_test_trap_subprocess(NULL, 0, 0);
+    g_test_trap_assert_failed();
+    g_test_trap_assert_stderr(msg);
+}
+
+static void test_sc_cgroupv2_own_group_path_die(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+    g_file_set_contents(fixture->p, (char *)user_data, -1, &err);
+    g_assert_no_error(err);
+    _test_sc_cgroupv2_own_group_path_die_with_message("unexpected content of group entry 0::\n");
+}
+
+static void test_sc_cgroupv2_own_group_path_no_file(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    g_remove(fixture->p);
+    _test_sc_cgroupv2_own_group_path_die_with_message("cannot open *\n");
+}
+
+static void test_sc_cgroupv2_own_group_path_permission(cgroupv2_own_group_fixture *fixture, gconstpointer user_data) {
+    if (geteuid() == 0) {
+        g_test_skip("the test will not work when running as root");
+        return;
+    }
+    int ret = g_chmod(fixture->p, 0000);
+    g_assert_cmpint(ret, ==, 0);
+    _test_sc_cgroupv2_own_group_path_die_with_message("cannot open *: Permission denied\n");
+}
+
+static void __attribute__((constructor)) init(void) {
```

Comment:
I would prefer if the fixture data was a bit more realistic with example corresponding to both the scope we use for apps, and the services case

Reply:
tweaked the tests to use more realistic dataset